File: constructs/pulumi/java/src/main/java/com/nautilus/infra/pulumi/policy/Tagging.java
package com.nautilus.infra.pulumi.policy;

import java.util.HashMap;
import java.util.Map;

/**
 * Required tagging policy.
 * Injects mandatory tags onto every Azure resource.
 * Developers never call this directly — components call it automatically.
 */
public final class Tagging {

    private Tagging() {}

    /**
     * Returns the mandatory tag set for all Azure resources.
     * Required tags always win — extra keys are merged underneath them.
     */
    public static Map<String, String> requiredTags(
            String project,
            String environment,
            Map<String, String> extra) {

        var tags = new HashMap<>(extra != null ? extra : Map.of());
        tags.put("managed_by",  "pulumi");
        tags.put("project",     project);
        tags.put("environment", environment);
        return Map.copyOf(tags);
    }

    public static Map<String, String> requiredTags(String project, String environment) {
        return requiredTags(project, environment, null);
    }
}

File: constructs/pulumi/python/nautilus_infra_pulumi/policy/tagging.py
"""
Required tagging policy
=======================
Injects mandatory tags onto every Azure resource.
Developers never call this directly — components call it automatically.
"""

from typing import Optional


def required_tags(project: str, environment: str, extra: Optional[dict] = None) -> dict:
    """Return the mandatory tag set for all Azure resources.

    Required tags always win — extra keys are merged underneath them.
    """
    tags: dict = {
        "managed_by":  "pulumi",
        "project":     project,
        "environment": environment,
    }
    if extra:
        return {**extra, **tags}
    return tags
